A62-pick-first.md

```diff
@@ -9,63 +9,68 @@ A62: `pick_first` LB policy
   before attempting to reconnect.
 
-Since `pick_first` was implemented prior to the first stable version of any gRPC
-language, no gRFC exists for it. Although it is a simple LB policy whose
-behavior can be summarized in a few sentences, as done above, there are some
-subtle differences in implementation across the various languages. This document
-aims to provide a more detailed description of its behavior, with the aim of
-bringing convergence among the different implementations.
+When connections to all addresses fail, there are some similarities and some
+differences between the Java/Go implementations and the C-core implementation.
 
-## Proposal
+Similarities include:
+- Reporting `TRANSIENT_FAILURE` as the connectivity state to the channel.
+- Applying exponential backoff. During the backoff period, `wait_for_ready` RPCs
+  are queued while other RPCs fail.
+
+Differences show up after the backoff period ends:
+- C-core remains in `TRANSIENT_FAILURE` while Java/Go move to `IDLE`.
+- C-core attempts to reconnect to the given addresses, while Java/Go rely on the
+  client application to make an RPC or an explicit attempt to connect.
+- C-core moves to `READY` only when a connection attempt succeeds.
 
-Specific scenarios are described in their own subsections below.
+This behavior of staying in `TRANSIENT_FAILURE` until it can report `READY` is
+called sticky-TransientFailure, or sticky-TF.
 
-### When connections to all addresses fail
+## Proposal
+
+Specific changes are described in their own subsections below.
 
-`pick_first` LB policy attempts to connect to the given addresses in order, and
-when none of the addresses are reachable, it must:
-- report `TransientFailure` as the connectivity state to the channel
-- apply exponential backoff (all RPCs attempted during this period will fail)
-- attempt to reconnect to the given addresses in order
-- continue to report `TransientFailure` as the connectivity state, until a
-  connection succeeds, at which point, it must report `Ready`
+### Support Sticky-TF
 
-This behavior of staying in `TransientFailure` until it can report `Ready` is
-called sticky-TransientFailure.
+All `pick_first` implementations should support sticky-TF. Once connections to
+all addresses fail, they should:
+- Report `TRANSIENT_FAILURE` as the connectivity state.
+- Attempt to reconnect to the addresses indefinitely until a connection succeeds
+  (at which point, they should report `READY`), or the channel becomes idle (see
+  [gRPC Connectivity Semantics](1) for more details about idleness).
 
-`pick_first` LB policy should indefinitely attempt to reconnect, in a bid to
-move out of `TransientFailure` and into `Ready`, until the channel enters idle
-mode due to inactivity, at which point, the channel will shut down the name
-resolver and the LB policy. See [gRPC Connectivity Semantics](1) for more
-details about idleness.
+Supporting sticky-TF has the following advantages:
```

Comment:
Currently, this first explains the change being proposed and then explains why.  I think it would be more readable if we did that in the opposite order: first we explain the problem we're seeing with the current behavior, and then we can specify the change we're proposing to fix it.

Same thing in the "Enable random shuffling of address list" section below.

Reply:
Done. Thanks.

Reply:
That's definitely better.  But I think Eric's right that it would make more sense to move the problem statement up to the "Background" section.  I suggest the following:

- On line 31, add a paragraph that says "There are a few problems with the existing `pick_first` functionality, which will be described in the following subsections."
- Right after that paragraph, add a sub-section for explaining the sticky-TF problem.  That can include the text from lines 31-46 that describes what sticky-TF is, and it can then immediately segue into the text from lines 63-84 that explain why the lack of sticky-TF is a problem.
- Right after that sub-section, add another sub-section to explain the address randomization problem.  This will replace the text currently on lines 103-109.  I suggest the following wording:
"""
Because `pick_first` sends all requests to the same address, it is often used for L4 load balancing by randomizing the order of the addresses used by each client.  In general, gRPC expects address ordering to be determined as part of name resolution, not by the LB policy.  For example, DNS servers may randomize the order of addresses when there are multiple A/AAAA records, and the DNS resolver in gRPC is expected to perform [RFC-6724 address sorting](https://www.rfc-editor.org/rfc/rfc6724).  However, there are some cases where DNS cannot randomize the address order, either because the DNS server does not support that functionality or because it is defeated by client-side DNS caching.  To address such cases, it is desirable to add a client-side mechanism for randomly shuffling the order of the addresses.
"""
- This discussion also made me realize that this gRFC should also include the fact that we're going to add the ability to configure `pick_first` via xDS.  To that end, I suggest another sub-section here that says something like this:
"""
There are cases where it is desirable to perform L4 load balancing using `pick_first` when getting addresses via xDS instead of DNS.  As a result, we need a way to configure use of this LB policy via xDS.
\
Note that client-side address shuffling may be equally desirable in this case, since the xDS server may send the same EDS resource (with the same endpoints in the same order) to all clients.
"""
- Add another sub-section under "Proposal" that describes the ability to configure `pick_first` via xDS.  This should reference [gRFC A52](https://github.com/grpc/proposal/blob/master/A52-xds-custom-lb-policies.md) and https://github.com/envoyproxy/envoy/pull/26952.

Reply:
Done. Thank you so much for the detailed suggestions.